File: src/discord/renderers/answerDelivery.ts
import type { ComponentRow, EditableMessage, MessageChannel, OutgoingMessage } from '../ports.js';
import type { ImageRenderer } from '../render/segment.js';
import { splitAnswerSegments } from '../render/blockParser.js';
import { chunkMessage } from '../format.js';

// Common answer-delivery helper shared by the three final-text paths (streaming
// finalize, Codex plainText, Claude result-only fallback). Sends the answer in ORDER;
// when an ImageRenderer is injected, GFM tables and ```mermaid``` fences become inline
// PNG attachments in place (text → image → text), otherwise it is byte-for-byte the old
// `chunkMessage` behavior (so existing tests are unaffected).
//
// `firstSink` lets the streaming path reuse its live "Responding…" embed message as the
// FIRST output (edited in place, interrupt button disabled); later outputs are fresh
// sends. Without a sink, everything is a plain send.

export interface DeliverOptions {
  channel: MessageChannel;
  // Absent → text-only (the render branch is off / Chrome unavailable). Present → render.
  renderImage?: ImageRenderer;
  // Streaming live embed to reuse for the first output (edit in place). null/undefined
  // → all outputs are fresh sends.
  firstSink?: EditableMessage | null;
  // Components (the disabled interrupt button) to leave on the finalized first message.
  disabledComponents?: ComponentRow[];
}

function rawTextForBlock(seg: { kind: 'table' | 'mermaid'; source?: string; code?: string }): string {
  return seg.kind === 'mermaid' ? '```mermaid\n' + (seg.code ?? '') + '\n```' : seg.source ?? '';
}

export async function deliverAnswer(text: string, opts: DeliverOptions): Promise<void> {
  const { channel, renderImage, disabledComponents } = opts;
  let sink = opts.firstSink ?? null;

  // Send one payload; the FIRST call reuses the live embed (edit in place, clearing the
  // embed and disabling the interrupt button), the rest are fresh sends. discord.js edit
  // accepts new `files`, so an image can replace the embed without deleting the message.
  const emit = async (payload: OutgoingMessage): Promise<void> => {
    if (sink) {
      const m = sink;
      sink = null;
      await m.edit({
        content: payload.content ?? '',
        embeds: [],
        ...(payload.files ? { files: payload.files } : {}),
        ...(disabledComponents ? { components: disabledComponents } : {}),
      });
      return;
    }
    await channel.send(payload);
  };

  // No renderer → preserve the exact legacy behavior: chunk the whole text, first chunk
  // into the sink (with the disabled button), the rest as follow-up sends.
  if (!renderImage) {
    for (const chunk of chunkMessage(text)) await emit({ content: chunk });
    // Nothing to send (empty answer) but a live embed lingers → clear it.
    if (sink) await sink.edit({ content: '​', embeds: [], components: [] }).catch(() => {});
    return;
  }

  const segs = splitAnswerSegments(text);
  for (const seg of segs) {
    if (seg.kind === 'text') {
      for (const chunk of chunkMessage(seg.text)) await emit({ content: chunk });
      continue;
    }
    const img = await renderImage.render(seg);
    if (img) {
      await emit({ files: [{ path: img.data, name: img.name }] });
    } else {
      // Render skipped/failed → keep the block's original markdown (answer never breaks).
      for (const chunk of chunkMessage(rawTextForBlock(seg))) await emit({ content: chunk });
    }
  }
  if (sink) await sink.edit({ content: '​', embeds: [], components: [] }).catch(() => {});
}

File: swift/Sources/DiscordAgentBridge/Render/AnswerDelivery.swift
import Foundation

// Common answer-delivery helper (TS `answerDelivery.ts`).
// When `renderImage` is set, GFM tables and ```mermaid``` fences become PNG attachments
// in place (text → image → text); otherwise byte-for-byte `DiscordText.chunkMessage`.

/// One outbound payload: plain text and/or a single file attachment.
public struct DeliverPayload: Sendable, Equatable {
    public var content: String?
    public var fileName: String?
    public var fileData: Data?

    public init(content: String? = nil, fileName: String? = nil, fileData: Data? = nil) {
        self.content = content
        self.fileName = fileName
        self.fileData = fileData
    }

    public static func text(_ s: String) -> DeliverPayload {
        DeliverPayload(content: s)
    }

    public static func file(name: String, data: Data) -> DeliverPayload {
        DeliverPayload(fileName: name, fileData: data)
    }
}

public struct DeliverOptions: Sendable {
    /// Present → render tables/mermaid; absent → text-only (legacy chunkMessage).
    public var renderImage: ImageRenderFn?
    /// Called once per payload, in order.
    public var emit: @Sendable (DeliverPayload) async throws -> Void
    /// When the answer produced nothing but a live sink still exists — clear it (TS sink.edit empty).
    public var clearEmpty: (@Sendable () async -> Void)?

    public init(
        renderImage: ImageRenderFn? = nil,
        emit: @escaping @Sendable (DeliverPayload) async throws -> Void,
        clearEmpty: (@Sendable () async -> Void)? = nil
    ) {
        self.renderImage = renderImage
        self.emit = emit
        self.clearEmpty = clearEmpty
    }
}

private func rawTextForBlock(_ seg: RenderableSegment) -> String {
    switch seg {
    case .mermaid(let code):
        return "```mermaid\n\(code)\n```"
    case .table(let source):
        return source
    }
}

/// Deliver answer text (optionally rendering table/mermaid blocks as PNGs).
/// Never throws from render failures — only `emit` / `clearEmpty` may throw.
public func deliverAnswer(_ text: String, options: DeliverOptions) async throws {
    let emit = options.emit
    let renderImage = options.renderImage

    // No renderer → preserve legacy: chunk whole text.
    guard let renderImage else {
        let chunks = DiscordText.chunkMessage(text)
        if chunks.isEmpty {
            await options.clearEmpty?()
            return
        }
        for chunk in chunks {
            try await emit(.text(chunk))
        }
        return
    }

    let segs = splitAnswerSegments(text)
    var emitted = false
    for seg in segs {
        switch seg {
        case .text(let t):
            for chunk in DiscordText.chunkMessage(t) {
                try await emit(.text(chunk))
                emitted = true
            }
        case .table, .mermaid:
            guard let block = RenderableSegment(seg) else { continue }
            if let img = await renderImage(block) {
                try await emit(.file(name: img.name, data: img.data))
                emitted = true
            } else {
                // Render skipped/failed → keep original markdown.
                for chunk in DiscordText.chunkMessage(rawTextForBlock(block)) {
                    try await emit(.text(chunk))
                    emitted = true
                }
            }
        }
    }
    if !emitted {
        await options.clearEmpty?()
    }
}
